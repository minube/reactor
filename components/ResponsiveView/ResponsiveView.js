import { func, node } from 'prop-types';
import React, { PureComponent } from 'react';
import { View } from 'react-native';

import { LAYOUT } from '../../common';

class ResponsiveView extends PureComponent {
  render() {
    const {
      children, onLayout, ...inherit
    } = this.props;

    return (
      <View
        {...inherit}
        key={this.state && this.state.viewport}
        onLayout={() => {
          const viewport = `${LAYOUT.VIEWPORT.W}x${LAYOUT.VIEWPORT.H}`;
          if (this.state && this.state.viewport === viewport) return;

          LAYOUT.calc();
          onLayout(LAYOUT.VIEWPORT);
          this.setState({ viewport });
          // this.forceUpdate(); @TODO: React fiber
        }}
      >
        {children}
      </View>
    );
  }
}

ResponsiveView.propTypes = {
  children: node,
  onLayout: func,
};

ResponsiveView.defaultProps = {
  children: undefined,
  onLayout() {},
};

export default ResponsiveView;
