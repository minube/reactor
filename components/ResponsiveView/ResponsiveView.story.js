import React from 'react';
import { View } from 'react-native';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withState } from '@dump247/storybook-state';

import ResponsiveView, { ResponsiveStyle } from './ResponsiveView';
import Text from '../Text';

storiesOf('✅ ResponsiveView', module)
  .addWithJSX('default', () => (
    <ResponsiveView />
  ))
  .addWithJSX('⚡ onLayout', () => (
    <ResponsiveView onLayout={action('ResponsiveView.onLayout()')} />
  ))
  .addWithJSX('🏀 Playground', withState({ timestamp: new Date() }, store => (
    <ResponsiveView {...store.state} onLayout={() => store.set({ timestamp: new Date() })}>
      <Text bold large>VIEWPORT</Text>

      <ResponsiveStyle>
        {({ viewport }) =>
          <View>
            <Text small>{`${viewport.W} x ${viewport.H}`}</Text>
            { viewport.PORTRAIT && <Text tiny>PORTRAIT</Text> }
            { viewport.LANDSCAPE && <Text tiny>LANDSCAPE</Text> }
            { viewport.IPHONEX && <Text tiny>IPHONEX</Text> }
            { viewport.TINY && <Text tiny>TINY</Text> }
            { viewport.PHONE && <Text tiny>PHONE</Text> }
            { viewport.TABLET && <Text tiny>TABLET</Text> }
            { viewport.SMALL && <Text tiny>SMALL</Text> }
            { viewport.REGULAR && <Text tiny>REGULAR</Text> }
            { viewport.LARGE && <Text tiny>LARGE</Text> }
          </View>
        }
      </ResponsiveStyle>
    </ResponsiveView>
  )));

