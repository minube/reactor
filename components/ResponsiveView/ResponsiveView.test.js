import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import ResponsiveView from './ResponsiveView';

describe('<ResponsiveView>', () => {
  it('renders', () => {
    const tree = renderer.create(<ResponsiveView />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {onLayout}', () => {
    // @TODO: Test <ResponsiveView /> {onLayout}
  });

  it('renders in DOM environment', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ResponsiveView />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
});

