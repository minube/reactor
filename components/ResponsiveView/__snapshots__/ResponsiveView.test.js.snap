// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResponsiveView> renders 1`] = `
<View
  onLayout={[Function]}
/>
`;
